LW-002

WojciechBarwinski opened the pull request:
utworzono 2 DTOs do przyjmowanie zamowienia oraz 1 dto do zwracania brakujących przedmiotów. Utworzono 1 controller który przyjmuje zamówienie

WojciechBarwinski committed:
LW-002
utworzono 2 DTOs do przyjmowanie zamowienia oraz 1 dto do zwracania brakujących przedmiotów.
Utworzono 1 controller który przyjmuje zamówienie

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/controllers/OrderController.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/lw")
```
/lw jest tu niepotrzebne - wiemy, że to jest aplikacja/usługa legendary warehouse

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/controllers/OrderController.java:
```diff
@@ -0,0 +1,42 @@
+public class OrderController {
+
+    @PostMapping("/order")
+    public ResponseEntity<?> correctOrder(@RequestBody OrderDTO order) {
```
Nie chcemy wildcarda

gestru replied:
Nie chcemy  wildcarda- w controllerach z reguły zakładamy tylko pozytywną ścieżkę, a zwracanie odpowiedzi z błędami robimy za pomocą ControllerAdvice/Exception Handlerów

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/controllers/OrderController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    private boolean checkIfOrderContainProductWithIDToIncorrectOrder(List<OrderLineDTO> orderLines) {
+        String incorrectProductID = "007";
```
To jest ewidentnie stała klasy, więc chcielibyśmy mieć ją jako stałą (pole private static final xxx: T)

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/controllers/OrderController.java:
```diff
@@ -0,0 +1,42 @@
+
+    private List<MissingProductDTO> createMissingProducts() {
+        List<MissingProductDTO> missingProductDTOList = new ArrayList<>();
+        missingProductDTOList.add(new MissingProductDTO("007", 3, 1));
```
Właśnie po to robimy stałe :D Gdybyś chciał zmienić to ide na inne musiałbyś to zrobić w dwóch miejscach

gestru replied:
Właśnie po to robimy stałe :D Gdybyś chciał zmienić to id na inne musiałbyś to zrobić w dwóch miejscach

WojciechBarwinski committed:
LW-002.2
Changes after review.
Create Exception and ExceptionHandler

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/controllers/ApplicationExceptionHandlerController.java:
```diff
@@ -0,0 +1,24 @@
+
+    @ResponseStatus(HttpStatus.UNPROCESSABLE_ENTITY) // 422
+    @ExceptionHandler(MissingProductException.class)
+    public List<MissingProductDTO> missingProductException(MissingProductException exception){
```
Nigdy nie zwracamy listy (generalnie kolekcji). Dobrą praktyką jest owrapowanie tego w jakiś obiekt (np. na przypadek, gdy musimy dorzucić jakieś metadane jak paginacja, total, czy link do kolejnej strony).
Tyczy to się nie tylko błędów, ale pozytywnych zwrotek z API również.

W kwestii obsługi wyjątków zaś dobrze jest zrobić jakąś klasę odpowiedzi z wyjątku (np. ErrorResponse<T>) i tam zrobić pole typu <T> na dane związane z wyjątkiem, a obok niego coś w stylu errorMessage

WojciechBarwinski replied:
Naprawione. jest w LW-002.3. Czy to tak miało być? Czy to poprawne użycie generyka?

WojciechBarwinski committed:
LW-002.3
Add ErrorResponse and change ExceptionHandler to return ErrorResponse

WojciechBarwinski reviewed (commented):
hmm... co sie stanie jak tu cos wpiesz?

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/exceptions/ErrorResponse.java:
```diff
@@ -0,0 +1,12 @@
+@Data
+@AllArgsConstructor
+public class ErrorResponse<T>{
+    private int errorCode;
```
Nie ma sensu zwracać tutaj takiego kodu - już mamy w samej odpowiedzi ;) 
A tak generujesz potencjalny rozjazd, bo co jak się developer walnie i będzie zwrócone HTTP 200, a tu będzie 422? :P
AAAle - możesz zrobić jak najbardziej customowe kody. Zwracając sam message dajesz małe pole do popisu klientom jak wyświetlić error ich clientom (np. na UI) - takie customowe kody są bardziej elastyczne, bo pozwalają podpiąć np. komunikat w różnych językach jak potrzeba.

W naszym wypadku:
 - albo wywalmy ten kod, message nam wystarczy
 - albo zamień to pole na Stringa i wrzuć tam jakiś customowy kod z nowego enuma
 -

WojciechBarwinski replied:
oki, wiec usunałem kod HTTP, ale do exceptiona dodałem pole message, wiec w miejscu rzucania błędu będzie podana wiadomość którą poźniej przekaże się do ErrorResponse. Czy tak może być? 
Zakładam ze prawie kazdy exception jakiąś wiadomośc powinien miec a w tym momencie ErrorResponse ją otrzymuje i z dowolnymi danymi przesyła dalej.

WojciechBarwinski committed:
LW-002.4
Remove invalid field from ErrorResponse and change MissingProductException. Now MissingProductException need message.

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/exceptions/MissingProductException.java:
```diff
@@ -0,0 +1,18 @@
+@Getter
+public class MissingProductException  extends ApplicationException{
+    private final List<MissingProductDTO> MISSING_PRODUCTS;
+    private final String MESSAGE;
```
message*

gestru commented on src/main/java/com/wojciechbarwinski/demo/legendary_warehouse/exceptions/MissingProductException.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+public class MissingProductException  extends ApplicationException{
+    private final List<MissingProductDTO> MISSING_PRODUCTS;
```
missingProducts

WojciechBarwinski committed:
LW-002.5
Change names 2 fields
